### ISSUE TITLE
Lint against #[no_mangle] with the rust ABI

### ISSUE THREAD

**Issue by bjorn3** [rust-lang/rust-clippy#10347]
### What it does

The lint should lint against `#[no_mangle]` on functions with the rust abi. See https://github.com/Rust-for-Linux/linux/pull/967 for an example of this kind of issue.

### Lint Name

no_mangle_with_rust_abi

### Advantage

The Rust ABI is not stable, but in many simple cases matches enough with the C ABI that it is possible to forget to add `extern "C"` to a function called from C. Changes to the Rust ABI can break this at any point.

### Example

```rust
#[no_mangle]
fn foo() {}
```

Should instead be written as:

```rust
#[no_mangle]
extern "C" fn foo() {}
```

**Comment by nindalf** [comment]
@bjorn3 I've implemented this is in a PR that's waiting on review. This is what the lint does

- iff the function has `no_mangle` AND `Abi == Abi::Rust`, suggest adding `extern "C"` to it.
- It will not run for any of the other variants defined [`rustc_target::spec::abi::Abi`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_target/spec/abi/enum.Abi.html), nor suggest any conversion other than conversion to the C ABI.

My question to you is - is it possible to have false positives on this lint, to your knowledge? Is there some possible use case for adding `#[no_mangle]` to a Rust ABI function?

**Comment by bjorn3** [comment]
I don't think there are many use cases. Maybe you could make it such that using an explicit `extern "Rust"` doesn't give a warning, but omitting `extern` entirely does? This way you can suppress the lint using `extern "Rust"` and at the same time indicate that this was what you actually meant.

**Comment by nindalf** [comment]
I've added an exception for `extern "Rust"`. Suppressing this lint `#[allow(no_mangle_with_rust_abi)]` on an item or the whole crate is another option users have. 

If we can't think of any use cases for using `#[no_mangle]` on a Rust ABI function, I'm leaning towards this lint being opt-out rather than opt-in.

Probably a call for the clippy maintainers to take. @llogiq does that seem fine?

**Comment by llogiq** [comment]
I'm not completely clear on the semantics here: Doesn't the ABI default to Rust, thus making no ABI specification equal to `extern "Rust"`? If that's right, I don't see the benefit of this lint in the general case, and it should probably be allow by default.

**Comment by bjorn3** [comment]
When using `#[no_mangle]` you almost definitively intended it to be called from C and thus must not use the Rust abi. If you did actually mean to use the Rust abi, writing an explicit `extern "Rust"` helps clarify this.

**Comment by ojeda** [comment]
> If we can't think of any use cases for using #[no_mangle] on a Rust ABI function

The `__rust_*` functions are one such case in Rust for Linux, but we had to declare them due to the fact that some crate types did not generate them -- https://github.com/Rust-for-Linux/linux/pull/967#issuecomment-1430056663.

**Comment by bjorn3** [comment]
Note the real versions of `__rg_*`, `__rdl_*` and `__rust_*` are actually `#[rustc_std_internal_symbol]` rather than `#[no_mangle]`. Both attributes are identical except that `#[rustc_std_internal_symbol]` isn't exported from cdylib's unlike `#[no_mangle]`. Also having to define those `__rust_*` functions will hopefully go away soon.

**Comment by nindalf** [comment]
Copy pasting what the current impl of the lint does. Let me know if this works for everyone

> This PR adds a new lint `no_mangle_with_rust_abi` that suggests converting a function to the C ABI to if the function has the `#[no_mangle]` attribute and `Abi == Abi::Rust`. It will not run for any of the other variants defined in [rustc_target::spec::abi::Abi](https://doc.rust-lang.org/beta/nightly-rustc/rustc_target/spec/abi/enum.Abi.html), nor suggest any conversion other than conversion to the C ABI.
>
> Functions that explicitly opt into the Rust ABI with `extern "Rust"` are ignored by this lint.

**Comment by ojeda** [comment]
> If that's right, I don't see the benefit of this lint in the general case, and it should probably be allow by default.

The problem is forgetting `extern "C"` when having written `#[no_mangle]`.

Perhaps somebody has a common enough use case for such combination, but it would seem that the lint should be warn or deny by default (not sure what is the threshold/policy for those in Clippy), especially if the explicit `extern "Rust"` escape hatch is provided.

That is, projects that really use it that way can either use the escape hatch or simply allow it, but everyone else should be informed about the (very likely) mistake they made.

**Comment by ojeda** [comment]
Is is possible to get a Crater run with the lint enabled as deny, by the way? It would be nice to know how common the combination is, how many things would break if it was deny by default, and to report/spot bugs in the wild if any.

**Comment by nindalf** [comment]
Not a crater run, but here are a couple of Github search queries

- [without `extern "C"`](https://github.com/search?q=language%3ARust+AND+%2F%23%5C%5Bno_mangle%5C%5D%5Cn%5Cw%2B%28pub%29%3F+fn+.*%2F&type=code&p=2) matches 4.5k times.
- [with `extern "C"`](https://github.com/search?q=language%3ARust+AND+%2F%23%5C%5Bno_mangle%5C%5D%5Cn%5Cw%2B%28pub%29%3F+extern+%22C%22+fn+.*%2F&type=code) matches 16.9k times.

@ojeda and @bjorn3 - are most of the examples from the first link making a mistake?

And follow up question - if they are making a mistake, @llogiq - is this an acceptable number of people impacted?

**Comment by nindalf** [comment]
Sorry about that, I've fixed the query now.

**Comment by bjorn3** [comment]
> `@ojeda` and `@bjorn3` - are most of the examples from the first link making a mistake?

I saw some cases of `#[no_mangle]` inside `extern "C" {}` blocks where it doesn't have any effect. Others indeed look like they are missing `extern "C"`.

**Comment by nindalf** [comment]
@llogiq what do you think? It seems like the lint is doing the right thing, but it will fire on a lot of code.

**Comment by nindalf** [comment]
One thing I want to add about the exemption for `extern "Rust"` - this works, but `rustfmt` overwrites by default. The rule `[force_explicit_abi](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#force_explicit_abi)` (on by default) will replace the extern declaration with an empty string ([src](https://github.com/rust-lang/rustfmt/blob/34f9ca28f2f4ae3ac462c6dcd862b513e107e459/src/utils.rs#L145-L146)). Therefore, we won't see this code unless the author skipped running `rustfmt`.

No change to the existing PR - the exemption is still there.

**Comment by bjorn3** [comment]
> but rustfmt overwrites it.

Didn't know that. That is somewhat annoying. At least `#[allow]` would still work.

**Comment by nindalf** [comment]
@bjorn3 and @ojeda - I'm trying to think of a way for kernel developers to use this lint without impacting other users. Would you be open to explicitly opting in with `#![warn(clippy:: no_mangle_with_rust_abi)]` in the kernel prelude? In that case we could change this lint to `pedantic` and merge it.

**Comment by ojeda** [comment]
Yeah, that is completely fine for us -- we do so already for other lints, via the command-line, see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Makefile#n479.

**Comment by nindalf** [comment]
The PR is merged.

It should be available in stable with Rust 1.69.0, expected April 20th. 🚀

**Comment by bjorn3** [comment]
Thanks for working on this!
